shaper.go

```diff
@@ -1,6 +1,141 @@
 package shaping
 
+import (
+	"fmt"
+
+	"github.com/benoitkugler/textlayout/harfbuzz"
+	"github.com/go-text/di"
+	"golang.org/x/image/font/sfnt"
+	"golang.org/x/image/math/fixed"
+)
+
 type Shaper interface {
 	// Shape takes an Input and shapes it into the Output.
 	Shape(Input) Output
 }
+
+// scale returns x divided by unitsPerEm, rounded to the nearest fixed.Int26_6
+// value (1/64th of a pixel). Borrowed from sfnt package.
+func scale(x fixed.Int26_6, unitsPerEm sfnt.Units) fixed.Int26_6 {
+	if x >= 0 {
+		x += fixed.Int26_6(unitsPerEm) / 2
+	} else {
+		x -= fixed.Int26_6(unitsPerEm) / 2
+	}
+	return x / fixed.Int26_6(unitsPerEm)
+}
+
+// scalePosition using specific ppem and upem values.
+func scalePosition(gp harfbuzz.GlyphPosition, ppem fixed.Int26_6, upem sfnt.Units) harfbuzz.GlyphPosition {
+	gp.XAdvance = int32(scale(fixed.I(int(gp.XAdvance)).Mul(ppem), upem).Round())
+	gp.YAdvance = int32(scale(fixed.I(int(gp.YAdvance)).Mul(ppem), upem).Round())
+	gp.XOffset = int32(scale(fixed.I(int(gp.XOffset)).Mul(ppem), upem).Round())
+	gp.YOffset = int32(scale(fixed.I(int(gp.YOffset)).Mul(ppem), upem).Round())
+	return gp
+}
+
+// TODO: can we just make Input and Output into structs?
+// Interfaces encapsulate varying behavior well, but these
+// types have no behavior. They're just structured data.
+// It feels silly to wrap them in accessors like this.
+type output struct {
+	DotAdvance    fixed.Int26_6
+	TopToBaseline fixed.Int26_6
+	TextBounds    fixed.Rectangle26_6
+	Glyphs        []Glyph
+}
+
+func (o output) Advance() fixed.Int26_6 {
+	return o.DotAdvance
+}
+
+func (o output) Baseline() fixed.Int26_6 {
+	return o.TopToBaseline
+}
+
+func (o output) Bounds() fixed.Rectangle26_6 {
+	return o.TextBounds
+}
+
+func (o output) Length() int {
+	return len(o.Glyphs)
+}
+
+func (o output) Index(i int) Glyph {
+	return o.Glyphs[i]
+}
+
+// Shape turns an input into an output.
+func Shape(input Input) Output {
+	// Prepare to shape the text.
+	// TODO: maybe reuse these buffers for performance?
```

Comment:
That is indeed a possible optimisation, since the `harfbuzz.Buffer` is designed to reuse allocated memory.
For instance, the Pango C library use a global variable to store one buffer (with a mutex) and reuse it.

Reply:
This would be interesting to explore down the road. Thanks!